tests/p2p/test_sync.py

```diff
@@ -305,6 +310,142 @@ def test_downloader(self):
+        # add blocks
+        blocks = self._add_new_blocks(10)
+        tx_A = blocks[0]
+        tx_B = blocks[1]
+
+        # create second peer
+        self.manager2 = self.create_peer(self.network)
+
+        # connect them and sync all blocks
+        conn0 = FakeConnection(self.manager1, self.manager2)
+        for _ in range(1000):
+            if conn0.is_empty():
+                break
+            conn0.run_one_step()
+            self.clock.advance(0.1)
+        else:
+            self.fail('expected to break out of loop')
+        self.assertTipsEqual(self.manager1, self.manager2)
+        self.assertEqual(self.manager1.tx_storage.get_best_block(), blocks[-1])
+        self.assertEqual(self.manager2.tx_storage.get_best_block(), blocks[-1])
+
+        # create the peer that will experience the bug
+        manager_bug = self.create_peer(self.network)
+        downloader = manager_bug.connections._sync_factories[SyncVersion.V1].downloader
+        downloader.window_size = 1
+        conn1 = FakeConnection(manager_bug, self.manager1)
+        conn2 = FakeConnection(manager_bug, self.manager2)
+
+        # put that peer in a situation where sync advanced to the point that tx_A and tx_B are requested
+        for _ in range(50):
+            conn1.run_one_step()
+            if tx_A.hash in downloader.pending_transactions and tx_B.hash in downloader.pending_transactions:
+                break
+        else:
+            self.fail('expected to break out of loop')
+
+        # force second download after clock is advanced
+        self.clock.advance(10.0)
+        downloader.start_next_download()
+
+        for _ in range(50):
+            conn2.run_one_step()
+            details_A = downloader.pending_transactions.get(tx_A.hash)
+            details_B = downloader.pending_transactions.get(tx_B.hash)
+            details_A_has_conns = details_A is not None and len(details_A.connections) >= 2
+            details_B_has_conns = details_B is not None and len(details_B.connections) >= 2
+            if details_A_has_conns and details_B_has_conns:
+                break
+        else:
+            self.fail('expected to break out of loop')
+
+        # by this point everything should be set to so we can trigger the bug, any issues that happen before this
+        # comment are an issue in setting up the scenario, not related to the problem itself
+
+        # ## triggering the issue
+
+        # wait for the download of tx_A to timeout but not yet the download of tx_B
+        conn1.disconnect(Failure(Exception('testing')))
+        conn2.disconnect(Failure(Exception('testing')))
+
+        # the previous behavior was to let the tx be removed only when the download request timeouts,
+        self.assertNotIn(tx_A.hash, downloader.pending_transactions)
+        self.assertNotIn(tx_B.hash, downloader.pending_transactions)
+
+        # XXX: the following sequence will trigger the bug when the solution is not being used, but since the behavior
```

Comment:
As far as I understood, if you comment out the code below, it would pass since the bug has been fixed. Is it correct? If yes, why don't you let it run?

Reply:
Not really, the sequence fails after the fix because we can't get `details_A` and `details_B`, because after the fix they are immediately removed when the peer disconnects. This could be worked around by not using `details_A`/`details_B` in the following `for`, however the timing for that `for` is somewhat sensitive, we need to advance it enough for the new connection to be added to the list of both tx_details, but not too much that it would download the transactions. The timing could be hardcoded in the number of steps because of the predictability, but if we do that, any change that causes a change in the timing of the sync would potentially make this test meaningless, it's not that it would break the test, but more that it wouldn't, so it would no longer reproduce the problem. Does that make sense?

Reply:
I've worked it around and made it so the base setup is a method and there are two tests, one that triggers the bug and checks whether the transaction was added successfully (which only happens when the bug is not triggered), and a second one that tests the change in behavior (that is, if a connection is immediately removed from tx-details when disconnected).

Reply:
Great job! Just reviewed the tests and approved the PR.